File: Source/OldCode/Archive/2005 TVA Code Library/Version/2.0/Local/Source/TVA.Phasors/HeaderFrameBase.vb
'*******************************************************************************************************
'  HeaderFrameBase.vb - Header frame base class
'  Copyright © 2005 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2005
'  Primary Developer: J. Ritchie Carroll, Operations Data Architecture [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  01/14/2005 - J. Ritchie Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

Imports System.Runtime.Serialization
Imports System.Text
Imports TVA.DateTime

Namespace Phasors

    ''' <summary>This class represents the protocol independent common implementation of a header frame that can be sent or received from a PMU.</summary>
    <CLSCompliant(False), Serializable()> _
    Public MustInherit Class HeaderFrameBase

        Inherits ChannelFrameBase(Of IHeaderCell)
        Implements IHeaderFrame

        Protected Sub New()
        End Sub

        Protected Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)

            MyBase.New(info, context)

        End Sub

        Protected Sub New(ByVal cells As HeaderCellCollection)

            MyBase.New(cells)

        End Sub

        ' Derived classes are expected to expose a Public Sub New(ByVal binaryImage As Byte(), ByVal startIndex As Int32)
        ' and automatically pass in parsing state
        Protected Sub New(ByVal state As IHeaderFrameParsingState, ByVal binaryImage As Byte(), ByVal startIndex As Int32)

            MyBase.New(state, binaryImage, startIndex)

        End Sub

        ' Derived classes are expected to expose a Public Sub New(ByVal headerFrame As IHeaderFrame)
        Protected Sub New(ByVal headerFrame As IHeaderFrame)

            MyClass.New(headerFrame.Cells)

        End Sub

        Protected Overrides ReadOnly Property FundamentalFrameType() As FundamentalFrameType
            Get
                Return TVA.Phasors.FundamentalFrameType.HeaderFrame
            End Get
        End Property

        Public Overridable Shadows ReadOnly Property Cells() As HeaderCellCollection Implements IHeaderFrame.Cells
            Get
                Return MyBase.Cells
            End Get
        End Property

        Public Overridable Property HeaderData() As String Implements IHeaderFrame.HeaderData
            Get
                Return Encoding.ASCII.GetString(Cells.BinaryImage)
            End Get
            Set(ByVal value As String)
                Cells.Clear()
                ParseBodyImage(New HeaderFrameParsingState(Cells, 0, value.Length), Encoding.ASCII.GetBytes(value), 0)
            End Set
        End Property

        Public Overrides ReadOnly Property Attributes() As Dictionary(Of String, String)
            Get
                Dim baseAttributes As Dictionary(Of String, String) = MyBase.Attributes

                baseAttributes.Add("Header Data", HeaderData)

                Return baseAttributes
            End Get
        End Property

    End Class

End Namespace

File: Source/OldCode/Archive/2007 Phasor Protocols/PhasorProtocols/HeaderFrameBase.cs
//*******************************************************************************************************
//  HeaderFrameBase.cs
//  Copyright © 2009 - TVA, all rights reserved - Gbtc
//
//  Build Environment: C#, Visual Studio 2008
//  Primary Developer: James R Carroll
//      Office: PSO TRAN & REL, CHATTANOOGA - MR BK-C
//       Phone: 423/751-4165
//       Email: jrcarrol@tva.gov
//
//  Code Modification History:
//  -----------------------------------------------------------------------------------------------------
//  01/14/2005 - James R Carroll
//       Generated original version of source code.
//
//*******************************************************************************************************

using System;
using System.Text;
using System.Collections.Generic;
using System.Runtime.Serialization;

namespace PCS.PhasorProtocols
{
    /// <summary>
    /// Represents the protocol independent common implementation of any <see cref="IHeaderFrame"/> that can be sent or received.
    /// </summary>
    [Serializable()]
    public abstract class HeaderFrameBase : ChannelFrameBase<IHeaderCell>, IHeaderFrame
    {
        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="HeaderFrameBase"/>.
        /// </summary>
        /// <param name="cells">The reference to the <see cref="HeaderCellCollection"/> for this <see cref="HeaderFrameBase"/>.</param>
        protected HeaderFrameBase(HeaderCellCollection cells)
            : base(0, cells, 0)
        {
        }

        /// <summary>
        /// Creates a new <see cref="HeaderFrameBase"/> from serialization parameters.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> with populated with data.</param>
        /// <param name="context">The source <see cref="StreamingContext"/> for this deserialization.</param>
        protected HeaderFrameBase(SerializationInfo info, StreamingContext context)
            : base(info, context)
        {
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets the <see cref="FundamentalFrameType"/> for this <see cref="HeaderFrameBase"/>.
        /// </summary>
        public override FundamentalFrameType FrameType
        {
            get
            {
                return FundamentalFrameType.HeaderFrame;
            }
        }

        /// <summary>
        /// Gets reference to the <see cref="HeaderCellCollection"/> for this <see cref="HeaderFrameBase"/>.
        /// </summary>
        public virtual new HeaderCellCollection Cells
        {
            get
            {
                return base.Cells as HeaderCellCollection;
            }
        }

        /// <summary>
        /// Gets or sets the parsing state for the this <see cref="HeaderFrameBase"/>.
        /// </summary>
        public virtual new IHeaderFrameParsingState State
        {
            get
            {
                return base.State as IHeaderFrameParsingState;
            }
            set
            {
                base.State = value;
            }
        }

        /// <summary>
        /// Gets or sets header data for this <see cref="HeaderFrameBase"/>.
        /// </summary>
        public virtual string HeaderData
        {
            get
            {
                return Encoding.ASCII.GetString(Cells.BinaryImage);
            }
            set
            {
                Cells.Clear();
                State = new HeaderFrameParsingState(0, value.Length);
                ParseBodyImage(Encoding.ASCII.GetBytes(value), 0, value.Length);
            }
        }

        /// <summary>
        /// <see cref="Dictionary{TKey,TValue}"/> of string based property names and values for the <see cref="HeaderFrameBase"/> object.
        /// </summary>
        public override Dictionary<string, string> Attributes
        {
            get
            {
                Dictionary<string, string> baseAttributes = base.Attributes;

                baseAttributes.Add("Header Data", HeaderData);

                return baseAttributes;
            }
        }

        #endregion
    }
}